4-storage/relational-database-shredder/src/main/scala/com.snowplowanalytics.snowplow.storage/spark/ShredJob.scala

```diff
@@ -0,0 +1,323 @@
+   * @return The original line with the proper fields removed respecting the Postgres constaints
+   */
+  def alterEnrichedEvent(originalLine: String, newEventId: Option[String]): String = {
+    import com.snowplowanalytics.snowplow.enrich.common.utils.ConversionUtils
+
+    // TODO: move PostgresConstraints code out into Postgres-specific shredder when ready.
+    // It's okay to apply Postgres constraints to events being loaded into Redshift as the PG
+    // constraints are typically more permissive, but Redshift will be protected by the
+    // COPY ... TRUNCATECOLUMNS.
+    val ignoredJsonFieldIndices = Set(52, 58, 122)
+    val eventIdIndex = 6
+    originalLine.split("\t", -1).toList.zipAll(utils.PostgresConstraints.maxFieldLengths, "", None)
+      .map { case (field, maxLength) =>
+        maxLength match {
+          case Some(ml) => ConversionUtils.truncate(field, ml)
+          case None => field
+        }
+      }
+      .zipWithIndex
+      .filter(x => !ignoredJsonFieldIndices.contains(x._2))
+      .map { case (value, index) =>
+        newEventId match {
+          case Some(eid) if index == eventIdIndex => eid
+          case _                                  => value
+        }
+      }
+      .mkString("\t")
+  }
+
+  /**
+   * Transform the shredded JSON hierarchy if the original event had synthetic duplicates.
+   * Inject a new event id if present as root id.
+   * @param hierarchy Shredded json hierarchy which may contain duplicated rood id
+   * @param newEventId Some event id if the original event was duplicated
+   * @return the original json if the event hadn't been duplicated, a new one with an updated root
+   *         id otherwise
+   */
+  def updateHierarchyWithDedupedId(hierarchy: JsonNode, newEventId: Option[String]): JsonNode = {
+    import org.json4s.jackson.JsonMethods.{fromJsonNode, asJsonNode}
+    import org.json4s.JObject
+    import org.json4s.JsonDSL._
+    newEventId
+      .map { id =>
+        val jsonObject = fromJsonNode(hierarchy).merge("hierarchy" -> ("rootId" -> id): JObject)
+        asJsonNode(jsonObject)
+      }
+      .getOrElse(hierarchy)
+  }
+
+  /**
+   * The path at which to store the altered enriched events.
+   * @param outFolder shredded/good/run=xxx
+   * @return The altered enriched event path
+   */
+  def getAlteredEnrichedOutputPath(outFolder: String): String = {
+    val alteredEnrichedEventSubdirectory = "atomic-events"
+    s"${outFolder}${if (outFolder.endsWith("/")) "" else "/"}${alteredEnrichedEventSubdirectory}"
+  }
+
+  /**
+   * The path at which to store the shredded types.
+   * @param outFolder shredded/good/run=xxx
+   * @return The shredded types output path
+   */
+  def getShreddedTypesOutputPath(outFolder: String): String = {
+    val shreddedTypesSubdirectory = "shredded-types"
+    s"${outFolder}${if (outFolder.endsWith("/")) "" else "/"}${shreddedTypesSubdirectory}"
+  }
+}
+
+/** Case class representing shredded events. */
```

Comment:
Can we use this opportunity to add @tparams for all of the case class properties